webpack/webpack-embedded-dependencies-plugin/src/EmbeddedDependenciesWebpackPlugin.ts

```diff
@@ -0,0 +1,361 @@
+   * you can call `plugin.apply(compiler)` to apply the plugin and invoke it.
+   * @param compiler - The webpack compiler instance.
+   */
+  public apply(compiler: Compiler): void {
+    // Tap into compilation so we can tap into compilation.hooks.processAssets
+    compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (compilation, { normalModuleFactory }) => {
+      const thirdPartyPackages: ThirdPartyPackageMap = new Map();
+
+      normalModuleFactory.hooks.module.tap(
+        PLUGIN_NAME,
+        (module, moduleCreateData: IWebpackModuleCreateData, resolveData) => {
+          const { resourceResolveData } = moduleCreateData;
+          const pkg: IPackageData | undefined = resourceResolveData?.descriptionFileData;
+          const filePath: string | undefined = resourceResolveData?.descriptionFileRoot;
+
+          if (pkg && filePath && filePath.includes('node_modules')) {
+            const key: PackageMapKey = makePackageMapKeyForPackage(pkg);
+            thirdPartyPackages.set(key, { dir: filePath, data: pkg });
+          }
+
+          return module;
+        }
+      );
+
+      compilation.hooks.processAssets.tapPromise(
+        { name: PLUGIN_NAME, stage: Compilation.PROCESS_ASSETS_STAGE_REPORT },
+        async (assets) => {
+          const packages: IPackageData[] = [];
+
+          Async.forEachAsync(Array.from(thirdPartyPackages), async ([, { dir, data }]) => {
+            const { name, version } = data;
+            let licenseSource: string | undefined;
+            const license: string | undefined = parseLicense(data);
+            const licensePath: string | undefined = await this._getLicenseFilePath(dir, compiler);
+            if (licensePath) {
+              licenseSource = await FileSystem.readFile(licensePath);
+
+              const copyright: string | undefined =
+                (await this._parseCopyright(licenseSource)) || parsePackageAuthor(data);
+
+              packages.push({
+                name,
+                version,
+                license,
+                licenseSource,
+                copyright
+              });
+            } else {
+              // If there is no license file path, we still should populate the other required fields
+              const copyright: string | undefined = parsePackageAuthor(data);
+
+              packages.push({
+                name,
+                version,
+                license,
+                copyright
+              });
+            }
+          })
+            .catch((error: unknown) => {
+              this._emitWebpackError(compilation, 'Failed to process embedded dependencies', error);
+            })
+            .finally(() => {
+              // Sort the packages by name after we have processed all of them
+              // Note this is a mutable sort
+              Sort.sortBy(packages, (pkg) => pkg.name);
+            });
```

Comment:
Put this in a normal `try`/`catch`/`finally` and await the `Async.forEachAsync` call?